False positive in unused_closure_parameter

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Bug Report

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint --path test.swift
Linting Swift files at path test.swift
Linting 'test.swift' (1/1)
/Users/jsimard/Downloads/test.swift:3:5: warning: Unused Closure Parameter Violation: Unused parameter "manager" in a closure should be replaced with _. (unused_closure_parameter)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?

0.24.0

* Installation method used (Homebrew, CocoaPods, building from source, etc)?

Homebrew

* Paste your configuration file:

None.

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

None.

* Which Xcode version are you using (check `xcode-select -p`)?

/Applications/Xcode-9.2.app/Contents/Developer

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
import Foundation

({ (manager: FileManager) in
  print(manager)
})(FileManager.default)
```

Also, as reported by @irace:

![image](https://user-images.githubusercontent.com/474794/34273368-067ed370-e662-11e7-8845-4546527a6260.png)

What customer service – thanks JP!

On Dec 21, 2017, 3:17 PM -0500, JP Simard <notifications@github.com>, wrote:
> New Issue Checklist
>
> • Updated SwiftLint to the latest version
> • I searched for existing GitHub issues
>
> Bug Report
> Complete output when running SwiftLint, including the stack trace and command used
> $ swiftlint lint --path test.swift
> Linting Swift files at path test.swift
> Linting 'test.swift' (1/1)
> /Users/jsimard/Downloads/test.swift:3:5: warning: Unused Closure Parameter Violation: Unused parameter "manager" in a closure should be replaced with _. (unused_closure_parameter)
> Done linting! Found 1 violation, 0 serious in 1 file.
> Environment
>
> • SwiftLint version (run swiftlint version to be sure)?
>
> 0.24.0
>
> • Installation method used (Homebrew, CocoaPods, building from source, etc)?
>
> Homebrew
>
> • Paste your configuration file:
>
> None.
>
> • Are you using nested configurations?
> If so, paste their relative paths and respective contents.
>
> None.
>
> • Which Xcode version are you using (check xcode-select -p)?
>
> /Applications/Xcode-9.2.app/Contents/Developer
>
> • Do you have a sample that shows the issue? Run echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules
> to quickly test if your example is really demonstrating the issue. If your example is more
> complex, you can use swiftlint lint --path [file here] --no-cache --enable-all-rules.
>
> import Foundation
>
> ({ (manager: FileManager) in
> print(manager)
> })(FileManager.default)
> Also, as reported by @irace:
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -32,6 +32,11 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#1968](https://github.com/realm/SwiftLint/issues/1968)
 
+* Fix false positives in `unused_closure_parameter` rule when closure is wrapped
+  in parentheses.  
+  [JP Simard](https://github.com/jpsim)
+  [#1979](https://github.com/realm/SwiftLint/issues/1979)
+
 ## 0.24.0: Timed Dry
 
 ##### Breaking

--- a/Rules.md
+++ b/Rules.md
@@ -15391,6 +15391,12 @@ hoge(arg: num) { num in
 
 ```
 
+```swift
+({ (manager: FileManager) in
+  print(manager)
+})(FileManager.default)
+```
+
 </details>
 <details>
 <summary>Triggering Examples</summary>

--- a/Source/SwiftLintFramework/Rules/UnusedClosureParameterRule.swift
+++ b/Source/SwiftLintFramework/Rules/UnusedClosureParameterRule.swift
@@ -41,7 +41,12 @@ public struct UnusedClosureParameterRule: ASTRule, ConfigurationProviderRule, Co
             "}(UILabel())\n",
             "hoge(arg: num) { num in\n" +
             "  return num\n" +
-            "}\n"
+            "}\n",
+            """
+            ({ (manager: FileManager) in
+              print(manager)
+            })(FileManager.default)
+            """
         ],
         triggeringExamples: [
             "[1, 2].map { ↓number in\n return 3\n}\n",
@@ -151,7 +156,7 @@ public struct UnusedClosureParameterRule: ASTRule, ConfigurationProviderRule, Co
         return dictionary.name.flatMap { name -> Bool in
             let length = name.bridge().length
             let range = NSRange(location: 0, length: length)
-            return regex("\\A\\s*\\{").firstMatch(in: name, options: [], range: range) != nil
+            return regex("\\A[\\s\\(]*?\\{").firstMatch(in: name, options: [], range: range) != nil
         } ?? false
     }
 

